src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,361 @@
+If `copycols=false`, then returned `DataFrame` shares column vectors with `df` where possible.
+
+If `df` is a `SubDataFrame` then a `SubDataFrame` is returned if `copycols=false`
+and a `DataFrame` with freshly allocated columns otherwise.
+
+Note that if `df` is a `DataFrame` and `copycols=false` then including the same column several times
+in the resulting data frame will create aliases.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select(df, :b)
+3×1 DataFrame
+│ Row │ b     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 4     │
+│ 2   │ 5     │
+│ 3   │ 6     │
+
+julia> select(df, Not(:b)) # drop column :b from df
+3×1 DataFrame
+│ Row │ a     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 1     │
+│ 2   │ 2     │
+│ 3   │ 3     │
+
+julia> select(df, :a=>:c, :b)
+3×2 DataFrame
+│ Row │ c     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select(df, :b, :a=><(1.5)=>:c)
+3×2 DataFrame
+│ Row │ b     │ c    │
+│     │ Int64 │ Bool │
+├─────┼───────┼──────┤
+│ 1   │ 4     │ 1    │
+│ 2   │ 5     │ 0    │
+│ 3   │ 6     │ 0    │
+```
+
+"""
+select(df::DataFrame, inds::AbstractVector{Int}; copycols::Bool=true) =
+    DataFrame(_columns(df)[inds], Index(_names(df)[inds]),
+              copycols=copycols)
+select(df::DataFrame, c::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                               Colon, All, Not, Between, Regex}; copycols::Bool=true) =
+    select(df, index(df)[c], copycols=copycols)
+select(df::DataFrame, c::ColumnIndex; copycols::Bool=true) =
+    select(df, [c], copycols=copycols)
+
+select(df::DataFrame, cs...; copycols::Bool=true) =
+    _select(df, [normalize_selection(index(df), c) for c in cs], copycols)
+
+function _select(df::DataFrame, ncs, copycols::Bool)
+    newdf = DataFrame()
+    # it should be OK to be type unstable here + in this way we aviod having to compile custom Dict
+    transformed_cols = Dict()
+    for nc in ncs
+        if nc isa Pair
+            newname = last(last(nc))
+            @assert newname isa Symbol
+            if haskey(transformed_cols, newname)
+                throw(ArgumentError("duplicate target transformed or renamed column names passed"))
+            end
+            transformed_cols[newname] = nc
+        end
+    end
+    for nc in ncs
+        if nc isa Union{Int, AbstractVector{Int}}
+            allunique(nc) || throw(ArgumentError("duplicate column names selected"))
+            for i in nc
+                newname = _names(df)[i]
+                if !hasproperty(newdf, newname)
+                    if haskey(transformed_cols, newname)
+                        nct = transformed_cols[newname]
+                        @assert !isnothing(nct)
```

Comment:
Inference likes this better AFAIK:
```suggestion
                        @assert nct !== nothing
```